### PR TITLE
Pal 972 resource redis ttl

### PR DESCRIPTION
**Pull request by developer01189998819991197253**

### CODE REVIEW

File: resource-service/src/component-tests/resources/application-testcontainers.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Substitute maven properties for testcontainers image names
+testcontainers:
+  redis.image: "@testcontainers.redis.image@"
+  redis.default.image: "redis:6-alpine"
+  kafka.image: "@testcontainers.kafka.image@"
+  kafka.default.image: "confluentinc/cp-kafka:5.5.1"
```
Comment (dev930018) [inline resource-service/src/component-tests/resources/application-testcontainers.yaml]:
```suggestion
```
not needed in component-tests

File: resource-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/resource/redis/RedisPersistenceTest.java [hunk 2]
```diff
@@ -60,11 +60,11 @@
 @SpringBootTest(
         classes = {RedisPersistenceTest.class, ResourceApplication.class},
         webEnvironment = WebEnvironment.RANDOM_PORT,
-        properties = {"akka.discovery.config.services.kafka.from-config=false", "spring.data.redis.repositories.key-prefix=test:"}
+        properties = {"akka.discovery.config.services.kafka.from-config=false"}
 )
 @Import({KafkaTestConfiguration.class})
 @ContextConfiguration(initializers = {RedisInitialiser.class})
-@ActiveProfiles({"redis", "akka-test", "testcontainers"})
+@ActiveProfiles({"redis-test", "akka-test", "testcontainers"})
```
Comment (dev930018) [inline resource-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/resource/redis/RedisPersistenceTest.java]: Revert these changes, i believe this is left over from debugging

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java [hunk 3]
```diff
@@ -36,9 +38,11 @@
...
 
 import java.lang.reflect.Field;
+import java.time.Duration;
 import java.util.Arrays;
 import java.util.LinkedList;
 import java.util.function.Function;
+import java.util.stream.Collectors;
 
 /**
  * A class that implements a {@link ReactiveCrudRepository} for Redis backing store
```
Comment (dev930018) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: Some more documentation may be needed here, just describing what the keys <-> values look like in redis and what (redis) types those values are

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java [hunk 4]
```diff
@@ -48,12 +52,22 @@
  */
 public abstract class AbstractReactiveRepositoryRedisAdapter<V, K> implements ReactiveCrudRepository<V, K> {
     public static final String KEY_SEP = "::";
+    private static final String ID_KEYSPACE = "id";
+    private static final String PARENT_KEYSPACE = "parent";
+    private static final String FORMAT_KEYSPACE = "format";
+    private static final String TYPE_KEYSPACE = "type";
```
Comment (dev930018) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: Each of these should be within resource, serialisedFormat and type classes respectively

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java [hunk 5]
```diff
@@ -222,44 +241,49 @@ protected AbstractReactiveRepositoryRedisAdapter(final ReactiveRedisTemplate<Str
...
+            return this.valueOps.delete(this.table + KEY_SEP + key)
+                    .then();
         }
     }
 
     /**
      * A class to allow {@link ResourceEntity}s to be stored in a reactive repository
      */
     public static class ResourceRepositoryAdapter extends AbstractReactiveRepositoryRedisAdapter<ResourceEntity, String> implements ResourceRepository {
+        private static final String SEPARATOR = KEY_SEP + PARENT_KEYSPACE + KEY_SEP;
```
Comment (dev930018) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: I'd rename this to maybe `PARENT_SEP` or similar

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java [hunk 6]
```diff
@@ -312,38 +339,40 @@ public SerialisedFormatRepositoryAdapter(final ReactiveRedisTemplate<String, Ser
...
+                    .map(entity -> this.setOps.remove(this.table + SEPARATOR + entity.getSerialisedFormat()))
+                    .then(this.deleteByIdDefault(key));
         }
     }
 
     /**
      * A class to allow {@link TypeRepository}s to be stored in a reactive repository
      */
     public static class TypeRepositoryAdapter extends AbstractReactiveRepositoryRedisAdapter<TypeEntity, String> implements TypeRepository {
+        private static final String SEPARATOR = KEY_SEP + TYPE_KEYSPACE + KEY_SEP;
```
Comment (dev930018) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: I'd rename this to `TYPE_SEP` or similar

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java [hunk 7]
```diff
@@ -270,38 +294,41 @@ public ResourceRepositoryAdapter(final ReactiveRedisTemplate<String, ResourceEnt
...
+                    .map(entity -> this.setOps.remove(this.table + SEPARATOR + entity.getParentId()))
+                    .then(this.deleteByIdDefault(key));
         }
     }
 
     /**
      * A class to allow {@link SerialisedFormatEntity}s to be stored in a reactive repository
      */
     public static class SerialisedFormatRepositoryAdapter extends AbstractReactiveRepositoryRedisAdapter<SerialisedFormatEntity, String> implements SerialisedFormatRepository {
+        private static final String SEPARATOR = KEY_SEP + FORMAT_KEYSPACE + KEY_SEP;
```
Comment (dev930018) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: I'd rename this to `SERIALISED_FORMAT_SEP` or similar

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java (hunk 4 above)
Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: Moved to the relevant sub classes

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java (hunk 5 above)
Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: Renamed

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java (hunk 6 above)
Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: Renamed

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java (hunk 7 above)
Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: Renamed

File: resource-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/resource/redis/RedisPersistenceTest.java (hunk 2 above)
Comment (pd104923) [reply resource-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/resource/redis/RedisPersistenceTest.java]: Reverted

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java (hunk 3 above)
Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: JavaDoc updated

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java [hunk 8]
```diff
@@ -36,24 +38,52 @@
...
 import java.util.LinkedList;
 import java.util.function.Function;
+import java.util.stream.Collectors;
 
 /**
- * A class that implements a {@link ReactiveCrudRepository} for Redis backing store
- *
+ * An abstract RedisAdapter class that implements a {@link ReactiveCrudRepository} for Redis backing store.
+ * The tables within the Redis backing store will have a unique key which is created using different elements, each element is separated by a double colon `::`.
+ * <p></p>
```
Comment (developer01189998819991197253) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: empty "< p >" tag

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java [hunk 9]
```diff
@@ -36,24 +38,52 @@
...
+ *     <tr>
+ *         <td>resources</td><td>|</td><td>Table Name :: parent :: Resource ID</td><td>|</td><td>Set of all the associated child resource IDs</td>
+ *     </tr>
+ *     <tr>
+ *         <td>types</td><td>|</td><td>Table Name ::type :: Type Value</td><td>|</td><td>Set of all resources associated with this type</td>
+ *     </tr>
+ *     <tr>
+ *         <td>serialised_formats</td><td>|</td><td>Table Name :: format :: Serialised Format Value</td><td>|</td><td>Set of all resources associated with this format</td>
+ *     </tr>
+ * </table>
```
Comment (developer01189998819991197253) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java:68]: no summary or caption for table

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java [hunk 10]
```diff
@@ -36,24 +38,52 @@
...
+ * <p></p>
```
Comment (developer01189998819991197253) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: empty `<p>` tag

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java (hunk 8 above)
Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: Resolved

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java (hunk 9 above)
Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java:68]: Resolved

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java (hunk 10 above)
Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/AbstractReactiveRepositoryRedisAdapter.java]: Resolved

File: resource-service/src/component-tests/java/uk/gov/gchq/palisade/component/resource/repository/RedisPersistenceLayerTest.java [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+
+    @Test
+    void testPersistenceTtlForExpiredInFlightResources() throws InterruptedException {
+        // When getting a resource from the persistence layer by resourceId, imitating a slow request
+        var inFlight = persistenceLayer.getResourcesById(resource.getId())
+                .join().orElseThrow();
+
+        // Then Sleep for 2 seconds to imitate a slow request
+        // The persistence evict for completeness happens here and also for resources
+        TimeUnit.SECONDS.sleep(15);
```
Comment (dev930018) [inline resource-service/src/component-tests/java/uk/gov/gchq/palisade/component/resource/repository/RedisPersistenceLayerTest.java]: While this has been _okay_ in previous tests, here we spend nearly a minute sleeping and waiting.
A better option might be to (either through reflection or extending the class in some way) set the TTL on a per-method basis, such that in each case we are waiting something like 1s for our expected ttl 'window', and the next longest expiry is suitably distant that we don't get issues.

For example, this would have a
```
// Given we have set the completeness AND resource repository ttl to 1s
// When we write resources to redis and then sleep for 1s
// Then the resources are evicted
```

**Review by dev930018 — CHANGES_REQUESTED** [review]

**Review by dev930018 — APPROVED** [review]
